fatal error LNK1179 on windows-x86_64-dmd with MSVC

## SHOO (@shoo) reported this on 2020-03-11T13:31:32Z

### Transferred from https://issues.dlang.org/show_bug.cgi?id=20667

### CC List

* 

### Description

````markdown
This code causes a link error(LNK1179):

foo/bar.d
-------------
import std.algorithm;
-------------

foo/main.d
-------------
module foo.main;
import std.path;
void foo(string file) { file.relativePath; }
-------------

hoge.d
-------------
import foo.main;
void main() {  }
-------------

Compile with follows:
-------------
dmd -m64 -lib -offoo.lib -cov foo\bar.d foo\main.d
dmd -m64 -cov hoge.d foo.lib
foo.lib(main.obj) : fatal error LNK1179: ファイルが無効か、または壊れています: COMDAT '__bzeroBytes' を複製します。(Maybe, in English: invalid or corrupt file: duplicate COMDAT '__bzeroBytes')
Error: linker exited with status 1179
-------------

Conditions: On Windows, x86_64(-m64), coverage(-cov) is enabled, and the installed MSVC(vs_buildtools 2019) environment.
It does not reproduce on other OS, MinGW(MSVC is not installed), x86, or without -cov.
And if removing unnecessary import statements or unnecessary files, the situation will not be reproduced.


Below is a similar link error but the situation seemed different.
https://issues.dlang.org/show_bug.cgi?id=20565
````

#### kntroh commented on 2021-04-29T06:38:04Z



````markdown
I got a similar error. However, the error code was LNK2005:

source/test.d:
---
import b;
void main() { }
---

sub/a.d:
---
import std.algorithm;
---

sub/b.d:
---
module b;
import std;
class C { void c() { f(); } }
struct S {
  double a = 0;
  string b;
  S[] c;
  @disable
  this ();
}
S f() {
  int[] arr;
  array(map!(a => f())(arr));
  assert (0);
}
---

dub.json:
---
{
  "name": "test",
  "targetType": "executable",
  "application": "test.d",
  "dependencies": { ":sub": "*" },
  "subPackages": [{
    "name": "sub",
    "targetType": "library",
    "sourcePaths": [ "sub" ],
    "importPaths": [ "sub" ]
  }]
}
---

When I prepare these files and build it by dub on Windows, I got an error that "__bzeroBytes already defined".
I have confirmed the reproduction with dmd 2.096.1 on Windows 8.1 and Windows 10 (20H2).

$ dub
Performing "debug" build using ***\dmd2\windows\bin\dmd.exe for x86_64.
test:sub ~master: target for configuration "library" is up to date.
test ~master: building configuration "application"...
Linking...
test_sub.lib(b_33_53.obj) : error LNK2005: __bzeroBytes は既に test_sub.lib(b_40_9af.obj) で定義されています。(__bzeroBytes already defined in test_sub.lib(h_48_8a7.obj))
.dub\build\application-debug-windows-x86_64-dmd_v2.096.1-dirty-4872E8CFA9A99F10C9963544DB0FC94E\test.exe : fatal error LNK1169: 1 つ以上の複数回定義されているシンボルが見つかりました。(found one or more symbols with multiple definitions)
Error: linker exited with status 1169
***\dmd2\windows\bin\dmd.exe failed with exit code 1.
$

I tried to see if I could make the reproduction code smaller, but it was impossible. For example, just replace `double a = 0;` with `double a = 1;` and it will not reproduce.
````

--- a/compiler/src/dmd/glue/toobj.d
+++ b/compiler/src/dmd/glue/toobj.d
@@ -396,9 +396,9 @@ void toObjFile(Dsymbol ds, bool multiobj)
 
                 // Generate static initializer
                 auto sinit = toInitializer(sd);
-                if (sinit.Sclass == SC.extern_)
+                if (sinit.Sclass == SC.extern_ &&
+                    strcmp(sinit.Sident.ptr, "__bzeroBytes") != 0) // might be from another object file
                 {
-                    if (sinit == bzeroSymbol) assert(0);
                     sinit.Sclass = sd.isInstantiated() ? SC.comdat : SC.global;
                     sinit.Sfl = FL.data;
                     auto dtb = DtBuilder(0);
@@ -589,8 +589,7 @@ void toObjFile(Dsymbol ds, bool multiobj)
                     scclass = SC.comdat;
 
                 // Generate static initializer
-                toInitializer(ed);
-                auto sinit = cast(Symbol*) ed.sinit;
+                auto sinit = toInitializer(ed);
                 sinit.Sclass = scclass;
                 sinit.Sfl = FL.data;
                 auto dtb = DtBuilder(0);

--- a/compiler/test/runnable/extra-files/gc19675.d
+++ b/compiler/test/runnable/extra-files/gc19675.d
@@ -1,0 +1,10 @@
+module gc19675;
+import memory19675;
+
+class ManualGC
+{
+    GC.ProfileStats profileStats()
+    {
+        return typeof(return).init;
+    }
+}

--- a/compiler/test/runnable/extra-files/main19675.d
+++ b/compiler/test/runnable/extra-files/main19675.d
@@ -1,0 +1,10 @@
+import gc19675;
+import memory19675;
+
+void main()
+{
+	auto p = new ManualGC;
+	auto s = p.profileStats();
+	auto s2 = GC.profileStats();
+	assert(s.numCollections[0] == 0);
+}

--- a/compiler/test/runnable/extra-files/memory19675.d
+++ b/compiler/test/runnable/extra-files/memory19675.d
@@ -1,0 +1,16 @@
+
+module memory19675;
+
+struct GC
+{
+    struct ProfileStats
+    {
+        size_t[10] numCollections;
+    }
+	static ProfileStats profileStats()
+    {
+        return typeof(return).init;
+    }
+
+}
+

--- a/compiler/test/runnable/test19675.sh
+++ b/compiler/test/runnable/test19675.sh
@@ -1,0 +1,6 @@
+#!/usr/bin/env bash
+
+$DMD -m${MODEL} -I${EXTRA_FILES} -of${OUTPUT_BASE}${LIBEXT} -lib ${EXTRA_FILES}/gc19675.d ${EXTRA_FILES}/memory19675.d
+$DMD -m${MODEL} -I${EXTRA_FILES} -of${OUTPUT_BASE}${EXE} ${EXTRA_FILES}/main19675.d ${OUTPUT_BASE}${LIBEXT}
+
+rm_retry ${OUTPUT_BASE}{${LIBEXT},${EXE}}

